### PR TITLE
ReadOnlySequence: Don't leak the flags that are on _startInteger and _endInteger.

### LINKED ISSUE #42999: When SequenceReader is completed, it's Position is not equal to it's Sequence.End.
### Description
```csharp
ReadOnlySequence<byte> buffer = new ReadOnlySequence<byte>( new byte[50] );
SequenceReader<byte> reader = new SequenceReader<byte>( buffer );
reader.Advance( 50 );
reader.End.Should().BeTrue();// Pass
buffer.Slice( reader.Position ).Length.Should().Be( 0 );// Pass
reader.Position.Equals( buffer.End ).Should().BeTrue();// Fail
```
I don't understand why this test fails, is this intended ?
The documentation tell 
> Equality does not guarantee that the two instances point to the same location in a ReadOnlySequence<T>.  

With this, I understand that, two instance not pointing to the same location can be equals,
but not that two SequencePositiong pointing to the same thing are not equal.

Discussed this with @tannergooding on discord.

### PR DESCRIPTION
**Pull request by Kuinox**
Fixes #42999.
Currently the ReadOnlySequence.End has sometimes a flag that causes odd behavior like:  
`GetPosition(size) != End`
`ReadOnlySequence<byte>.Empty.Start != ReadOnlySequence<byte>.Empty.End`  
This causes more odd behavior later on types relying on the ROSequence.

My fix simply removes the flags when we return a SequencePosition in .Start and .End.

### CODE REVIEW

File: src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs
Comment (Treit) [inline src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs:664]: Small typo / grammar: should probably be "Flag that allows encoding..."

Comment (Kuinox) [reply src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs]: Fixed.

**Comment by Kuinox** [comment]
And here the benchmarks:  
```
master:
|                          Method |      Mean |     Error |    StdDev |    Median |       Min |       Max 
|-------------------------------- |----------:|----------:|----------:|----------:|----------:|----------:
|              IterateTryGetArray |  6.292 ns | 0.0532 ns | 0.0471 ns |  6.273 ns |  6.255 ns |  6.408 ns 
|             IterateForEachArray | 17.566 ns | 0.1753 ns | 0.1640 ns | 17.485 ns | 17.393 ns | 17.747 ns 
|         IterateGetPositionArray | 17.354 ns | 0.1607 ns | 0.1503 ns | 17.349 ns | 17.155 ns | 17.589 ns 
|                      FirstArray |  6.720 ns | 0.1333 ns | 0.2036 ns |  6.737 ns |  6.353 ns |  7.212 ns 
|                  FirstSpanArray |  8.822 ns | 0.1050 ns | 0.0983 ns |  8.786 ns |  8.677 ns |  9.001 ns 
|                 FirstSpanMemory | 11.549 ns | 0.2215 ns | 0.2072 ns | 11.593 ns | 11.202 ns | 11.832 ns 
|          FirstSpanSingleSegment | 18.710 ns | 0.0468 ns | 0.0365 ns | 18.721 ns | 18.661 ns | 18.784 ns 
|            FirstSpanTenSegments | 19.025 ns | 0.1547 ns | 0.1447 ns | 18.955 ns | 18.897 ns | 19.265 ns 
|                      SliceArray |  6.021 ns | 0.0539 ns | 0.0478 ns |  6.034 ns |  5.939 ns |  6.078 ns 
|             IterateTryGetMemory | 24.220 ns | 0.2423 ns | 0.2267 ns | 24.129 ns | 23.900 ns | 24.572 ns 
|            IterateForEachMemory | 33.338 ns | 0.3058 ns | 0.2860 ns | 33.289 ns | 32.997 ns | 33.735 ns 
|        IterateGetPositionMemory | 36.789 ns | 0.4317 ns | 0.4038 ns | 36.501 ns | 36.438 ns | 37.560 ns 
|                     FirstMemory |  7.687 ns | 0.1543 ns | 0.1585 ns |  7.621 ns |  7.490 ns |  7.929 ns 
|                     SliceMemory |  7.946 ns | 0.0651 ns | 0.0609 ns |  7.939 ns |  7.878 ns |  8.045 ns 
|      IterateTryGetSingleSegment | 13.065 ns | 0.1347 ns | 0.1260 ns | 13.051 ns | 12.875 ns | 13.207 ns 
|     IterateForEachSingleSegment | 24.571 ns | 0.2268 ns | 0.2121 ns | 24.497 ns | 24.339 ns | 24.875 ns 
| IterateGetPositionSingleSegment | 23.719 ns | 0.0713 ns | 0.0596 ns | 23.718 ns | 23.598 ns | 23.840 ns 
|              FirstSingleSegment |  4.408 ns | 0.0896 ns | 0.1066 ns |  4.363 ns |  4.305 ns |  4.699 ns 
|              SliceSingleSegment |  6.551 ns | 0.0118 ns | 0.0092 ns |  6.551 ns |  6.533 ns |  6.565 ns 
|        IterateTryGetTenSegments | 39.891 ns | 0.0123 ns | 0.0103 ns | 39.890 ns | 39.875 ns | 39.913 ns 
|       IterateForEachTenSegments | 60.201 ns | 0.3256 ns | 0.3045 ns | 60.111 ns | 59.876 ns | 60.760 ns 
|   IterateGetPositionTenSegments | 77.265 ns | 0.1021 ns | 0.0797 ns | 77.263 ns | 77.107 ns | 77.444 ns 
|                FirstTenSegments |  3.796 ns | 0.0249 ns | 0.0221 ns |  3.801 ns |  3.752 ns |  3.826 ns 
|                SliceTenSegments | 23.177 ns | 0.2675 ns | 0.2234 ns | 23.098 ns | 22.917 ns | 23.662 ns 

PR:  
|                          Method |      Mean |     Error |    StdDev |    Median |       Min |       Max 
|-------------------------------- |----------:|----------:|----------:|----------:|----------:|----------:
|              IterateTryGetArray |  6.242 ns | 0.1057 ns | 0.0937 ns |  6.245 ns |  6.069 ns |  6.368 ns 
|             IterateForEachArray | 17.536 ns | 0.2349 ns | 0.2082 ns | 17.436 ns | 17.378 ns | 18.034 ns 
|         IterateGetPositionArray | 16.901 ns | 0.3560 ns | 0.3957 ns | 16.662 ns | 16.314 ns | 17.492 ns 
|                      FirstArray |  7.023 ns | 0.2879 ns | 0.8444 ns |  6.662 ns |  6.230 ns |  8.954 ns 
|                  FirstSpanArray |  8.850 ns | 0.0681 ns | 0.0637 ns |  8.844 ns |  8.725 ns |  8.942 ns 
|                 FirstSpanMemory | 11.418 ns | 0.2213 ns | 0.2070 ns | 11.402 ns | 10.959 ns | 11.801 ns 
|          FirstSpanSingleSegment | 18.867 ns | 0.1446 ns | 0.1353 ns | 18.790 ns | 18.741 ns | 19.162 ns 
|            FirstSpanTenSegments | 19.079 ns | 0.2027 ns | 0.1896 ns | 18.989 ns | 18.880 ns | 19.387 ns 
|                      SliceArray |  5.985 ns | 0.0697 ns | 0.0652 ns |  5.955 ns |  5.916 ns |  6.134 ns 
|             IterateTryGetMemory | 24.326 ns | 0.5120 ns | 1.3930 ns | 24.035 ns | 22.405 ns | 27.494 ns 
|            IterateForEachMemory | 34.008 ns | 0.7100 ns | 1.0843 ns | 33.843 ns | 32.760 ns | 36.556 ns 
|        IterateGetPositionMemory | 37.690 ns | 0.6607 ns | 0.6180 ns | 38.062 ns | 36.653 ns | 38.565 ns 
|                     FirstMemory |  7.775 ns | 0.1560 ns | 0.2564 ns |  7.630 ns |  7.411 ns |  8.207 ns 
|                     SliceMemory |  8.066 ns | 0.1601 ns | 0.1572 ns |  8.027 ns |  7.872 ns |  8.299 ns 
|      IterateTryGetSingleSegment | 13.360 ns | 0.2759 ns | 0.2581 ns | 13.266 ns | 13.059 ns | 13.731 ns 
|     IterateForEachSingleSegment | 25.066 ns | 0.5198 ns | 0.5105 ns | 25.008 ns | 24.391 ns | 25.715 ns 
| IterateGetPositionSingleSegment | 24.344 ns | 0.2918 ns | 0.2729 ns | 24.542 ns | 23.905 ns | 24.632 ns 
|              FirstSingleSegment |  4.395 ns | 0.0400 ns | 0.0445 ns |  4.389 ns |  4.297 ns |  4.497 ns 
|              SliceSingleSegment |  6.606 ns | 0.0852 ns | 0.0797 ns |  6.594 ns |  6.523 ns |  6.714 ns 
|        IterateTryGetTenSegments | 43.273 ns | 0.8911 ns | 1.4133 ns | 42.584 ns | 41.741 ns | 46.095 ns 
|       IterateForEachTenSegments | 60.702 ns | 0.8259 ns | 0.7726 ns | 60.471 ns | 59.866 ns | 61.716 ns 
|   IterateGetPositionTenSegments | 78.647 ns | 1.2567 ns | 1.1755 ns | 78.457 ns | 77.200 ns | 80.417 ns 
|                FirstTenSegments |  4.037 ns | 0.0831 ns | 0.2100 ns |  3.951 ns |  3.728 ns |  4.398 ns 
|                SliceTenSegments | 23.666 ns | 0.4604 ns | 0.6603 ns | 23.374 ns | 23.042 ns | 25.128 ns 
```

**Comment by Kuinox** [comment]
Currently, the ROSequence methods that take an argument SequencePosition, do a bitmask every time to remove the flag.
Now that the flags are not leaked anymore, it allows removing most of the bit masking operations that were used to remove the bit flag on input data.  
So I tried it and it adds some slight improvement compared to master:  
```
|                          Method |      Mean |     Error |    StdDev |    Median |       Min |       Max | Code Size |
|-------------------------------- |----------:|----------:|----------:|----------:|----------:|----------:|----------:|
|              IterateTryGetArray |  6.744 ns | 0.1788 ns | 0.5272 ns |  6.481 ns |  6.172 ns |  8.062 ns |     677 B |
|             IterateForEachArray | 17.309 ns | 0.1689 ns | 0.1580 ns | 17.236 ns | 17.132 ns | 17.569 ns |     859 B |
|         IterateGetPositionArray | 16.235 ns | 0.1603 ns | 0.1499 ns | 16.233 ns | 15.967 ns | 16.475 ns |     576 B |
|                      FirstArray |  6.354 ns | 0.0975 ns | 0.0912 ns |  6.325 ns |  6.196 ns |  6.490 ns |   4,011 B |
|                  FirstSpanArray |  8.503 ns | 0.1180 ns | 0.1104 ns |  8.471 ns |  8.366 ns |  8.716 ns |   8,791 B |
|                 FirstSpanMemory |  9.394 ns | 0.1900 ns | 0.4516 ns |  9.420 ns |  8.517 ns | 10.182 ns |   9,146 B |
|          FirstSpanSingleSegment | 18.967 ns | 0.0547 ns | 0.0427 ns | 18.963 ns | 18.912 ns | 19.061 ns |   8,939 B |
|            FirstSpanTenSegments | 18.942 ns | 0.1987 ns | 0.1859 ns | 18.841 ns | 18.643 ns | 19.173 ns |   8,939 B |
|                      SliceArray |  5.795 ns | 0.0462 ns | 0.0432 ns |  5.816 ns |  5.735 ns |  5.849 ns |   1,702 B |
|             IterateTryGetMemory | 23.421 ns | 0.3964 ns | 0.3310 ns | 23.424 ns | 22.958 ns | 24.024 ns |   1,032 B |
|            IterateForEachMemory | 32.435 ns | 0.5633 ns | 0.5269 ns | 32.101 ns | 31.799 ns | 33.276 ns |   1,214 B |
|        IterateGetPositionMemory | 36.889 ns | 0.2846 ns | 0.2662 ns | 36.738 ns | 36.477 ns | 37.191 ns |     931 B |
|                     FirstMemory |  7.799 ns | 0.1570 ns | 0.2201 ns |  7.697 ns |  7.527 ns |  8.137 ns |   4,366 B |
|                     SliceMemory |  7.921 ns | 0.0845 ns | 0.0749 ns |  7.888 ns |  7.822 ns |  8.038 ns |   2,057 B |
|      IterateTryGetSingleSegment | 13.142 ns | 0.1510 ns | 0.1413 ns | 13.231 ns | 12.768 ns | 13.255 ns |     825 B |
|     IterateForEachSingleSegment | 24.620 ns | 0.2728 ns | 0.2552 ns | 24.671 ns | 24.203 ns | 24.925 ns |   1,007 B |
| IterateGetPositionSingleSegment | 23.290 ns | 0.1898 ns | 0.1776 ns | 23.252 ns | 23.107 ns | 23.542 ns |     724 B |
|              FirstSingleSegment |  4.413 ns | 0.0751 ns | 0.0586 ns |  4.406 ns |  4.324 ns |  4.507 ns |   4,159 B |
|              SliceSingleSegment |  6.450 ns | 0.0543 ns | 0.0508 ns |  6.448 ns |  6.398 ns |  6.521 ns |   1,850 B |
|        IterateTryGetTenSegments | 37.278 ns | 0.3848 ns | 0.3599 ns | 37.018 ns | 36.988 ns | 38.050 ns |     825 B |
|       IterateForEachTenSegments | 60.194 ns | 0.5696 ns | 0.5049 ns | 60.261 ns | 59.592 ns | 60.860 ns |   1,007 B |
|   IterateGetPositionTenSegments | 73.585 ns | 0.5578 ns | 0.5218 ns | 73.424 ns | 72.657 ns | 74.303 ns |     724 B |
|                FirstTenSegments |  3.858 ns | 0.0647 ns | 0.0605 ns |  3.850 ns |  3.770 ns |  3.948 ns |   4,159 B |
|                SliceTenSegments | 22.181 ns | 0.1251 ns | 0.1045 ns | 22.208 ns | 21.983 ns | 22.353 ns |   1,850 B |
```
Do I have to open a new PR for it, or I append it to the current PR ?  
Note:
Currently this PR slow down a bit `IterateGetPositionTenSegments` (+1.4ns) because some bound checking is added in the JITed asm.  
Removing the bit masking on arguments removes the JITed bound checking.

File: src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs
Comment (adamsitnik) [inline src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs:65]: we can reuse the helper method that already takes care of that and is used in other places in the source code:

https://github.com/dotnet/runtime/blob/ec1897724f6dc5ab49eebde5f35292493e1bed7b/src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs#L467-L468

```suggestion
            get => new SequencePosition(_startObject, GetIndex(_startInteger));
```

File: src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs
Comment (adamsitnik) [inline src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.cs:74]: same as above:

```suggestion
            get => new SequencePosition(_endObject, GetIndex(_endInteger));
```

File: src/libraries/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.byte.cs
Comment (adamsitnik) [inline src/libraries/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.byte.cs:184]: could you please add a similar test for `Start`? And one for all positions from `Start+1` to `End - 1` (inclusive) that verifies that they are not equal to `.Start` and `.End`?

**Review by adamsitnik — COMMENTED** [review]
@Kuinox First of all big thanks for digging into this issue and finding the actual cause of the problem!

Overall looks good to me, but please add two more tests before we hit the merge button.

**Comment by Kuinox** [comment]
Adding the test you proposed showed a bug:  
`GetPosition(0) == buffer.Start`
failed: `Seek(0)` skipped the first buffer because it's empty.

I added a simple condition to stay on the SingleSegment Seek if the offset is 0.

No new benchmark because I currently have issues running the benchmark.
The change that I mentioned [there](https://github.com/dotnet/runtime/pull/47969#issuecomment-775545427), should I open a separate PR when this is closed, or I add it on top of this PR?

**Review by adamsitnik — APPROVED** [review]
LGTM, once again thank you @Kuinox !

File: src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs
Comment (stephentoub) [inline src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs:251]: Why is this change needed?

Comment (Kuinox) [reply src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs:251]: When Seek was called with the offset 0, it used SeekMultiSegment.
SeekMultiSegment will loop until it finds a segment with data, then return a SequencePosition with this segment.
So we ended up with a SequencePosition that point to the first nonempty segment, while `.Start` point to the first Segment.

Comment (stephentoub) [reply src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs:251]: Why is this specific to offset == 0? It's not the case any time currentLength == offset?

Comment (Kuinox) [reply src/libraries/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs:251]: When `currentLength == offset` it point to the first element of the next segment, not the current.
Except when it's 0, because we want to point to the start of the sequence.
But I just noticed that because of this, it's difficult to have SequencePosition pointing to the first byte of the Sequence.

**Pull request merged**
